src/main/java/uk/gov/service/notify/NotificationClient.java

```diff
@@ -215,14 +215,85 @@ public NotificationList getNotifications(String status, String notification_type
+     *
+     * @param templateId The template id is visible from the template page in the application.
+     * @return <code>Template</code>
+     */
+    public Template getTemplateById(String templateId) throws NotificationClientException{
+        String url = baseUrl + "/v2/template/" + templateId;
+        HttpsURLConnection conn = createConnectionAndSetHeaders(url, "GET");
+        String response = performGetRequest(conn);
+        return new Template(response);
+    }
+
+    /**
+     * The getTemplateVersion returns a <code>Template</code> given the template id and version.
+     *
+     *
+     * @param templateId The template id is visible from the template page in the application.
+     * @param version The version of the template to return
+     * @return <code>Template</code>
+     * @throws NotificationClientException
+     */
+    public Template getTemplateVersion(String templateId, int version) throws NotificationClientException{
+        String url = baseUrl + "/v2/template/" + templateId + "/version/" + version;
+        HttpsURLConnection conn = createConnectionAndSetHeaders(url, "GET");
+        String response = performGetRequest(conn);
+        return new Template(response);
+    }
+
+    /**
+     * Returns all the templates for your service. Filtered by template type if not null.
+     *
+     * @param templateType If templateType is not empty or null templates will be filtered by type.
+     *          Possible template types are email|sms|letter
+     * @return <code>TemplateList</code>
+     * @throws NotificationClientException
+     */
+    public TemplateList getAllTemplates(String templateType) throws NotificationClientException{
+        try{
+            URIBuilder builder = new URIBuilder(baseUrl + "/v2/templates");
+            if (templateType != null && !templateType.isEmpty()) {
+                builder.addParameter("type", templateType);
+            }
+            HttpsURLConnection conn = createConnectionAndSetHeaders(builder.toString(), "GET");
+            String response = performGetRequest(conn);
+            return new TemplateList(response);
+        } catch (URISyntaxException e) {
+            LOGGER.log(Level.SEVERE, e.toString(), e);
+            throw new NotificationClientException(e);
+        }
+    }
+
+    /**
+     * The getTemplatePreview returns a template with the placeholders replaced with the given personalisation.
+     *
+     * @param templateId The template id is visible from the template page in the application.
+     * @param personalisation Map representing the placeholders for the template if any. For example, key=name value=Bob
+     *                        Can be an empty map or null when the template does not require placeholders.
+     * @return <code>TemplatePreview</code>
+     * @throws NotificationClientException
+     */
+    public TemplatePreview getTemplatePreview(String templateId, Map<String, String> personalisation) throws NotificationClientException {
```

Comment:
I'd rather keep the method name getTemplatePreview or I'd change it to generateTemplatePreview